### PR TITLE
Ranking Evaluation for recommenders

### PR DESCRIPTION
**Pull request by dciborow**

### CODE REVIEW

File: src/core/test/fuzzing/src/test/scala/FuzzingTest.scala [hunk 1]
```diff
@@ -26,6 +26,10 @@ class FuzzingTest extends TestBase {
 
   test("Verify stage fitting and transforming") {
     val exemptions: Set[String] = Set(
+      "org.apache.spark.ml.tuning.RankingTrainValidationSplit",
```
Comment (mhamilton723) [inline src/core/test/fuzzing/src/test/scala/FuzzingTest.scala]: Lets hook into fuzzing for this, shouldnt be too hard

File: src/recommendation/src/main/python/RankingSplitters.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+# Licensed under the MIT License. See LICENSE in the project root for information.
+
+import numpy as np
+import pyspark
+
+# Methods for train-test split.
+class RankingSplitters:
+
+    @staticmethod
+    def min_rating_filter(self, min_rating, by_customer=True):
```
Comment (mhamilton723) [inline src/recommendation/src/main/python/RankingSplitters.py]: Can we put these on the scala side?

File: src/recommendation/src/main/python/RankingTrainValidationSplit.py [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+        return self._set(collectSubModels=value)
+
+    def getCollectSubModels(self):
+        """
+        Gets the value of collectSubModels or its default value.
+        """
+        return self.getOrDefault(self.collectSubModels)
+
+@inherit_doc
+class RankingTrainValidationSplit(Estimator, ValidatorParams, HasCollectSubModels, HasCollectSubMetrics,
```
Comment (mhamilton723) [inline src/recommendation/src/main/python/RankingTrainValidationSplit.py]: If this is an estimator it should be autogenned

File: src/recommendation/src/main/python/RankingTrainValidationSplit.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+from pyspark.ml.param import Params, Param, TypeConverters
+from pyspark.ml.tuning import ValidatorParams
+from pyspark.ml.util import *
+from pyspark.ml.param.shared import HasParallelism
+
+if sys.version >= '3':
+    basestring = str
+
+
+def _parallelFitTasks(est, train, eva, validation, epm, collectSubModel):
```
Comment (mhamilton723) [inline src/recommendation/src/main/python/RankingTrainValidationSplit.py]: Do you still need this function?

File: src/recommendation/src/main/scala/RankingHelper.scala [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    val est = $(estimator)
+    for (paramMap <- $(estimatorParamMaps).tail) {
+      est.copy(paramMap).transformSchema(schema)
+    }
+    est.copy(firstEstimatorParamMap).transformSchema(schema)
+  }
+
+}
+
+object RankingHelper {
```
Comment (mhamilton723) [inline src/recommendation/src/main/scala/RankingHelper.scala]: Do we  still use this

File: src/recommendation/src/test/python/RankingTrainValidationSplitSpec.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+            (3, 5, 3, 3),
+            (3, 6, 4, 4),
+            (3, 7, 5, 5),
+            (3, 8, 1, 1),
+            (3, 9, 5, 5),
+            (3, 10, 3, 3)], cSchema)
+        return ratings
+
+    def test_all_tiny(self):
+        os.environ["PYSPARK_DRIVER_PYTHON"] = "/home/dciborow/bin/python3"
```
[the remark this answers is not in the document]
Comment (dciborow) [reply src/recommendation/src/test/python/RankingTrainValidationSplitSpec.py]: moved to my bashrc, forgot to remove

File: src/core/test/fuzzing/src/test/scala/FuzzingTest.scala [hunk 7]
```diff
@@ -26,6 +26,9 @@ class FuzzingTest extends TestBase {
 
   test("Verify stage fitting and transforming") {
     val exemptions: Set[String] = Set(
+      "com.microsoft.ml.spark.RankingAdapterModel",
```
Comment (mhamilton723) [inline src/core/test/fuzzing/src/test/scala/FuzzingTest.scala]: Are these still needed?

File: src/recommendation/src/test/scala/RankingTrainValidationSplitSpec.scala [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+
+    val model = rankingTrainValidationSplit.fit(transformedDf)
+
+    val items = model.recommendForAllUsers(3)
+    val users = model.recommendForAllItems(3)
+    model.subMetrics.foreach(println(_))
+  }
+
+  override def testObjects(): Seq[TestObject[RankingTrainValidationSplit]] = {
+    val userCol = "customerIDOrg"
```
Comment (mhamilton723) [inline src/recommendation/src/test/scala/RankingTrainValidationSplitSpec.scala]: Can you share code between the first test and the the rest object? just pull the objects you need out of the first test and make them lazy

**Pull request closed without merging**